### PR TITLE
[HttpFoundation] Send cookies using header() to fix "SameSite" ones

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25344
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 1]
```diff
@@ -337,7 +337,7 @@ public function sendHeaders()
         }
 
         // headers
-        foreach ($this->headers->allPreserveCaseWithoutCookies() as $name => $values) {
+        foreach ($this->headers->allPreserveCase() as $name => $values) {
```
Comment (ThomHurks) [inline src/Symfony/Component/HttpFoundation/Response.php]: The current cookie code distinguishes between urlencoding the cookie value and not doing that (raw), the ```setcookie()``` function also performs some filtering and escaping as I understand it. As far as I can see in the code, the ```allPreserveCase()``` function doesn't do that but just casts the ```Cookie``` object to ```string``` and as such isn't a 1-to-1 substitute for ```setcookie()``` and ```setrawcookie()```, right?

**Comment by sroze** [comment]
Nop, the Cookie now understands how to stringify itself differently based
the raw or not raw. Therefore, we don't need PHP's `set*cookie`

**Comment by ThomHurks** [comment]
@sroze Yes, I realized that after examining the code more closely and so I edited my comment. Still, cross-referencing the code with PHP's implementation may be a good thing to do to prevent regression in behaviour. I'm not sure how new the Cookie code is, and the code that is currently there is not actually used to generate the final cookie at the moment, so it may be incomplete/untested?

**Comment by nicolas-grekas** [comment]
@ThomHurks here is the native implem:
https://github.com/php/php-src/blob/master/ext/standard/head.c#L84

everything looks fine to me.

**Comment by ThomHurks** [comment]
@nicolas-grekas Not the case, there are some differences that I can spot:
- Symfony does not check the value of ```value``` against the characters ```=,; \t\r\n\013\014``` whereas PHP does. Symfony only checks it against ```name``` and PHP checks it agains ```name``` and ```value```.
- When a cookie needs to be deleted, Symfony sets the ```max-age``` to ```31536001``` and the ```expiry``` to ```now - max-age``` but PHP just passes in ```expiry = 1``` and ```max-age = 0```.
- PHP checks if the expiry date has a year greater than 9999.
- Symfony also ```urlencode```'s ```name```, I don't see PHP do that.
- Symfony uses ```urlencode``` for the ```name``` vs ```rawurlencode``` for the ```value```, not sure why that difference is there.

**Comment by sroze** [comment]
I understand your concerns, and agree a bit. We will add integration
testing to this PR very soon. Highlighting the important scenarios that you
believe needs to be tested would be a very nice input from you! Could you
do so?

**Comment by nicolas-grekas** [comment]
I'd like to propose a plan:
- @sroze would you like to submit a testing infra for this? Could be heavily inspired by `AbstractSessionHandlerTest` on 3.4
- @ThomHurks would you like to submit a separate PR on 3.3 to align the behavior with native PHP?

**Comment by sroze** [comment]
I like this plan 👍

**Comment by ThomHurks** [comment]
@nicolas-grekas I will likely have time tomorrow to create a PR.

**Comment by nicolas-grekas** [comment]
@sroze @ThomHurks any progress?

**Comment by halaei** [comment]
Regarding [spectre attack](https://meltdownattack.com/) and [suggested mitigations](https://www.chromium.org/Home/chromium-security/ssca) should this be considered as a critical security fix?

**Comment by sroze** [comment]
I don't think it has anything to do with the attacks you've mentioned. I did not get time to get ahead with this so if anybody would like until I get back to it, you're welcome 😉

**Comment by halaei** [comment]
@sroze I quote the part from https://www.chromium.org/Home/chromium-security/ssca that seems to imply this is actually relevant:

> Where possible, prevent cookies from entering the renderer process' memory by using the SameSite and HTTPOnly cookie attributes, and by avoiding reading from document.cookie.

**Comment by xabbuh** [comment]
moving to the 3.4 milestone as the last bugfix release for 3.3 was published today

**Comment by nicolas-grekas** [comment]
Back on here, this PR now contains a second commit provided by @cvilleger, thank to him for working on it.

The Response + Cookie classes are now functionnally tested, and their behavior is a bit more aligned to PHP native one:

> Symfony does not check the value of value against the characters =,; \t\r\n\013\014 whereas PHP does. Symfony only checks it against name and PHP checks it agains name and value.

It does for the name, but doesn't care for the value. HTTP allows anything so no reason to restrict like PHP.

> When a cookie needs to be deleted, Symfony sets the max-age to 31536001 and the expiry to now - max-age but PHP just passes in expiry = 1 and max-age = 0.

Aligned

> PHP checks if the expiry date has a year greater than 9999.

I don't see any reason, kept as is (and tested.

> Symfony also urlencode's name, I don't see PHP do that.

PHP should encode. Symfony does, that's OK.

> Symfony uses urlencode for the name vs rawurlencode for the value, not sure why that difference is there.

Unrelated to the PHP vs Symfony delta, kept as is.

Status: needs review

File: src/Symfony/Component/HttpFoundation/Tests/CookieTest.php [hunk 2]
```diff
@@ -162,13 +162,13 @@ public function testCookieIsCleared()
     public function testToString()
     {
         $cookie = new Cookie('foo', 'bar', $expire = strtotime('Fri, 20-May-2011 15:25:52 GMT'), '/', '.myfoodomain.com', true);
-        $this->assertEquals('foo=bar; expires=Fri, 20-May-2011 15:25:52 GMT; max-age='.($expire - time()).'; path=/; domain=.myfoodomain.com; secure; httponly', (string) $cookie, '->__toString() returns string representation of the cookie');
+        $this->assertEquals('foo=bar; expires=Fri, 20-May-2011 15:25:52 GMT; Max-Age=0; path=/; domain=.myfoodomain.com; secure; httponly', (string) $cookie, '->__toString() returns string representation of the cookie');
```
Comment (phansys) [inline src/Symfony/Component/HttpFoundation/Tests/CookieTest.php]: Could you please update `assertEquals()` calls to `assertSame()`? (as done at `ClientTest`).

**Review by phansys — CHANGES_REQUESTED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by reatang** [comment]
Added the RFC document：rfc6265 
The Set-Cookie Header：
https://tools.ietf.org/html/rfc6265#section-5.2.2

**Comment by Seldaek** [comment]
@nicolas-grekas you said:

>> Symfony also urlencode's name, I don't see PHP do that.
> PHP should encode. Symfony does, that's OK.

This might be correct from a spec point of view, but FYI upgrading our app broke cookies with names "foo:bar:baz" as the : gets encoded and CloudFront's cookie whitelist does not recognize the encoded `foo%3Abar` as being the same cookie as `foo:bar` so the cookie gets dropped before reaching our servers. We could work around it but I wonder if this should be considered a regression or not.

**Comment by Seldaek** [comment]
Also a related issue, clearCookie does not allow clearing cookies without urlencoding, so a cookie set without urlencoding can not be cleared.

**Comment by nicolas-grekas** [comment]
@Seldaek I think we can fix these. WOuld you mind opening an issue?

**Pull request merged**